### PR TITLE
Kindergeldübertrag

### LINKED ISSUE #742: BUG: Kindergeldübertrag
### Bug description

The current version of GETTSIM doesn't support the Kindergeldübertrag.

The Kindergeldübertrag is relevant whenever children are not part of the Bedarfsgemeinschaft of their parents (e.g. due to enough income through Unterhalt). Then, the Kindergeld that is not necessary to cover the needs of the child are considered as parental income for Bürgergeld.

### PR DESCRIPTION
**Pull request by MImmesberger**
### What problem do you want to solve?

Closes #742 

The implementation will still rely on manually removing children from their Bedarfsgemeinschaft (via eigenbedarf_gedeckt) until the corresponding issue #622 has been addressed.

(This PR is based on #744)

### CODE REVIEW

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+"""Module for the calculation of the Kindergeldübertrag."""
```
Comment (MImmesberger) [inline src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:1]: For some reason, using the correct Umlaut in the filename results in gibberish and I get a ruff error in the pre-commits (N999)

Comment (hmgaudecker) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:1]: I think we are commonly doing it that way.

File: src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2_eink.py [hunk 2]
```diff
@@ -2,7 +2,42 @@
 from _gettsim.shared import policy_info
 
 
-def arbeitsl_geld_2_eink_m(  # noqa: PLR0913
+def arbeitsl_geld_2_eink_m(
+    _arbeitsl_geld_2_nettoeink_ohne_transfers_m: float,
+    kind_unterh_erhalt_m: float,
+    unterhaltsvors_m: float,
+    kindergeld_zur_bedarfdeckung_m: float,
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2_eink.py:9]:
```suggestion
    kindergeld_zur_bedarfsdeckung_m: float,
```
Or did they get rid of the Genitiv in official documents? 😉

File: src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2.py [hunk 3]
```diff
@@ -319,15 +317,10 @@ def arbeitsl_geld_2_vor_vorrang_m_bg(  # noqa: PLR0913
...
-    kindergeld_m_bg
-        See :func:`kindergeld_m_bg`.
-    kind_unterh_erhalt_m_bg
-        See basic input variable
-        :ref:`kind_unterh_erhalt_m_bg <kind_unterh_erhalt_m_bg>`.
-    unterhaltsvors_m_bg
-        See :func:`unterhaltsvors_m_bg`.
     arbeitsl_geld_2_eink_m_bg
         See :func:`arbeitsl_geld_2_eink_m_bg`.
+    kindergeld_übertrag_m_bg
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2.py:322]:
```suggestion
    kindergeldübertrag_m_bg
```
One word, no abbreviation = no underscore.

File: src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2_eink.py [hunk 4]
```diff
@@ -2,7 +2,42 @@
...
+) -> float:
+    """SGB II income.
+
+    Relevant income according to SGB II. Does not include child benefit transfer
+    (Kindergeldübertrag).
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2_eink.py:14]: Can you explain why? Intuitively, Kindergeldübertrag seems like a component of income to me, so I don't quite understand the structure of not including it there but deducting it from the payments in the end.

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+import numpy as np
+
+from _gettsim.shared import join_numpy, policy_info
+
+aggregate_by_p_id_kindergeldübertrag = {
+    "kindergeld_übertrag_m": {
+        "p_id_to_aggregate_by": "p_id_kindergeld_empf",
+        "source_col": "_kindergeld_kindbedarf_differenz_m",
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:10]:
```suggestion
        "source_col": "_diff_kindergeld_kindbedarf_m",
```
Easier to read, I'd say?

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        "aggr": "sum",
+    },
+}
+
+
+def _kindergeld_per_child_m(
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:16]:
```suggestion
def _mean_kindergeld_per_child_m(
```
I think we had that convention? 

And may I suggest separate functions for -2022, 2023+, mirroring the structure we have for Kindergeld?

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    kindergeld_m
+        See :func:`kindergeld_m`.
+    kindergeld_anz_ansprüche
+        See :func:`kindergeld_anz_ansprüche`.
+
+    Returns
+    -------
+
+    """
+    if kindergeld_anz_ansprüche == 0:
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:36]: Can we ever end up here? Why?

Put differently, this seems like doubling up what is already happening via `join_numpy` below?

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    which the child is entitled to is the sum of the Kindergeld for all children divided
+    by the amount of children. Hence, the age of the child (in comparison to siblings)
+    does not matter.
+
+    Parameters
+    ----------
+    kindergeld_m
+        See :func:`kindergeld_m`.
+    kindergeld_anz_ansprüche
+        See :func:`kindergeld_anz_ansprüche`.
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:63]: Remove

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    kindergeld_zur_bedarfdeckung_m: float,
+    eigenbedarf_gedeckt: bool,
+) -> float:
+    """Kindergeld that is used to cover the needs (SGB II) of the parent.
+
+    If a child does not need all of the Kindergeld to cover its own needs (SGB II), the
+    remaining Kindergeld is used to cover the needs of the parent (§ 11 Abs. 1 Satz 5
+    SGB II).
+
+
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:93]: Add something like: `kindergeldübetrag` obtains by aggregating this function at the parent level.

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
+"""Module for the calculation of the Kindergeldübertrag."""
+
+import numpy as np
+
+from _gettsim.shared import join_numpy, policy_info
+
+aggregate_by_p_id_kindergeldübertrag = {
+    "kindergeld_übertrag_m": {
+        "p_id_to_aggregate_by": "p_id_kindergeld_empf",
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: Just thinking of a weird case here: what if Kindergeldempfänger is outside the BG? E.g. ex-partner. Is that even allowed? Do we have/need a check?

File: src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      - true
+      - true
+      - true
+    bruttolohn_m:
+      - 20000.0
+      - 20000.0
+      - 0.0
+      - 0.0
+      - 0.0
+    _zu_verst_eink_ohne_kinderfreib_y_sn:
```
Comment (hmgaudecker) [inline src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml:55]: I thought `sn` were only the parents, so should be zero for the kids?

File: src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      - 200000.0
+  assumed: {}
+outputs:
+  kindergeld_m:
+    - 663.0
+    - 0.0
+    - 0.0
+    - 0.0
+    - 0.0
+  _kindergeld_erstes_kind_m:
```
Comment (hmgaudecker) [inline src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml:69]: Why would this be the same for all members of the household?

**Review by hmgaudecker — COMMENTED** [review]
Great work!

**Comment by hmgaudecker** [comment]
Is that skipped test related to #696? I thought the ALG2 comment there referred to other adults (who could work in principle), but not underage children?

**Comment by MImmesberger** [comment]
Yes, that's the bug we discovered some time ago. The skipped test shows that the `erwachsene_alle_rentner_hh` logic also fails for children because now they have a claim on the transfer themselves (I left a comment in the issue). 

We could fix this with a `and not kind` condition in `grunds_im_alter_m_eg`, or address the `erwachsene_alle_rentner_hh` when we fix #696.

File: src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml (hunk 12 above)
Comment (MImmesberger) [reply src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml:69]: Because this is a parameter that is the same for any individual given the policy date (as, for example, `minijob_grenze`)

File: src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2_eink.py (hunk 4 above)
Comment (MImmesberger) [reply src/_gettsim/transfers/arbeitsl_geld_2/arbeitsl_geld_2_eink.py:14]: I was leaning towards not counting Kindergeldübertrag as income of the parent (but of the child) but now that I'm thinking about it, I cannot defend that point.

Will switch to something like

```python
def arbeitsl_geld_2_eink_m(...):
   return arbeitsl_geld_2_eink_ohne_kindergeldübertrag_m + kindergeldübertrag_m
```

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py (hunk 7 above)
Comment (MImmesberger) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:36]: Yes because first `_mean_kindergeld_per_child_m` is being calculated (for any individual, so also for those without any children) and then the whole column is used by the `join_numpy`.
That's also what the tests say when removing that condition.

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py (hunk 10 above)
Comment (MImmesberger) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: The concept of Kindergeldübertrag is that the Kindergeldempfänger is in a different BG than the child (and hence there is Kindergeldübertrag). Are you thinking about different `fg`s?

I.e.
| p_id | hh_id | fg_id | bg_id | p_id_kindergeld_empf | p_id_elternteil_1 | p_id_elternteil_2 | kind  |
|------|-------|-------|-------|----------------------|-------------------|-------------------|-------|
| 0    | 0     | 0     | 0     | -1                   | -1                | -1                | false |
| 1    | 0     | 1     | 1     | -1                   | -1                | -1                | false |
| 2    | 0     | 0     | 2     | 1                    | 0                 | 1                 | true  |

I'd argue that this is not possible because then, the child should count towards the `fg` of `p_id==1`. But we don't have a test for that, we don't account for separated couples in the same household when creating `fg_id`.

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py (hunk 7 above)
Comment (hmgaudecker) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:36]: Cool, then maybe just switch the condition to go for `> 0` first, makes it easier to read (first what you care about, else other stuff)

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py (hunk 10 above)
Comment (hmgaudecker) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: Yes, I meant fg's.

Precondition for fg is living in the same household. I just wonder whether it is possible that the other parent gets the Kindergeld. And if not (what I'd kinda expect), whether we check that as a condition on the data.

**Comment by hmgaudecker** [comment]
> Yes, that's the bug we discovered some time ago. The skipped test shows that the `erwachsene_alle_rentner_hh` logic also fails for children because now they have a claim on the transfer themselves (I left a comment in the issue). 
> 
> We could fix this with a `and not kind` condition in `grunds_im_alter_m_eg`, or address the `erwachsene_alle_rentner_hh` when we fix #696.

Cool, pick whatever is easier.

File: src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml (hunk 12 above)
Comment (hmgaudecker) [reply src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml:69]: Is it worth testing that as an output then?

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py (hunk 10 above)
Comment (MImmesberger) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: I'll create an issue on the creation of `fg_id`. Intuitively, the child should form an `fg` with the parent that gets the Kindergeld (but could be something else of course). It's not clear to me what GETTSIM is currently doing in that regard.

Comment (hmgaudecker) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: I am pretty sure the fg conditions on being part of the same household. My main question is whether Kindergeld does the same when parents are separated.

Comment (MImmesberger) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: Kindergeld does not affect `fg_id` currently

Comment (hmgaudecker) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: Let me rephrase, that was not clear: 

My main question is whether `kindergeldempfänger` conditions that parent being part of the same household when parents are separated — and we are talking about means-tested transfers (ofc parents receive kindergeld for kids at college even though they may not be in the same household).

*(I am just worried that the machinery here depends implicitly on this assumption being true, would like to make it explicit)*

File: src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml (hunk 12 above)
Comment (MImmesberger) [reply src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml:69]: Wanted to test it somewhere, but making it part of an "economic" test case is probably not great. Should I move that to `test_kindergeld.py`?

Comment (hmgaudecker) [reply src/_gettsim_tests/test_data/kindergeld/2021/3_children.yaml:69]: I would be fine with testing it implicitly, but whatever you prefer!

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py (hunk 10 above)
Comment (MImmesberger) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: Okay so my point was that the creation of `fg_id` in case parents have a child, are separated **but live in the same household** should react to who claims Kindergeld.

Your point is that if parents **live in different households** `p_id_kindergeldempfänger` should be compatible with the place of living of the child (if such a requirement exists, presumably it does) if one of the involved individuals gets transfers?

> (I am just worried that the machinery here depends implicitly on this assumption being true, would like to make it explicit)

We create neither of the variables endogenously. I'm not sure to which extent creating a consistency check here would create a circle in the DAG.

Comment (hmgaudecker) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: > Your point is that if parents **live in different households** `p_id_kindergeldempfänger` should be compatible with the place of living of the child (if such a requirement exists, presumably it does) if one of the involved individuals gets transfers?

Exactly.

> > (I am just worried that the machinery here depends implicitly on this assumption being true, would like to make it explicit)
> 
> We create neither of the variables endogenously. I'm not sure to which extent creating a consistency check here would create a circle in the DAG.

Would be a check pre-DAG,  just whether the data passed conforms to expectations.

Comment (MImmesberger) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: But how do we check pre-DAG whether the involved individuals get transfers?

If I understood it correctly, if all involved individuals don't have their SGB II Bedarf gedeckt, we want to allow for

| p_id | hh_id | fg_id | bg_id | p_id_kindergeld_empfänger | kind  |
|------|-------|-------|-------|---------------------------|-------|
| 0    | 0     | 0     | 0     | -1                        | false |
| 1    | 0     | 0     | 0     | 0                         | true  |
| 2    | 1     | 1     | 1     | -1                        | false |

but we don't want to allow for / it would be inconsistent if

| p_id | hh_id | fg_id | bg_id | p_id_kindergeld_empfänger | kind  |
|------|-------|-------|-------|---------------------------|-------|
| 0    | 0     | 0     | 0     | -1                        | false |
| 1    | 0     | 0     | 0     | 2                         | true  |
| 2    | 1     | 1     | 1     | -1                        | false |

But if all adults have their SGB II Bedarf gedeckt, the latter situation would be perfectly fine.

Comment (hmgaudecker) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:9]: Fine, let's not get carried away here. It may be more about `alleinerziehend` status (or, more precisely, having `unterhaltsanspruch`) and `kindergeld`, basically it is baked in there already (deduction of 1/2 `kindergeld` from the aluminum payment). We should still think of a check for that, but no need to do it here. 

As always, closing an issue by thinking through it opens at least three new ones... 😂

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+    """
+    return kindergeld_params["kindergeld"] if kindergeld_anz_ansprüche > 0 else 0.0
+
+
+@policy_info(skip_vectorization=True)
+def kindergeld_zur_bedarfsdeckung_m(
+    _mean_kindergeld_per_child_m: float,
+    p_id_kindergeld_empf: np.ndarray[int],
+    p_id: np.ndarray[int],
+) -> float:
```
Comment (hmgaudecker) [inline src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:73]: Returns an array of floats, no?

**Review by hmgaudecker — APPROVED** [review]
Excellent! No show-stoppers left AFAICT!

File: src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py (hunk 13 above)
Comment (MImmesberger) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py]: I get the following error in the `_convert_data_to_correct_types` when using `numpy.ndarray[float]`:

```text
ValueError: The data types of the following columns are invalid:
E
E            - kindergeld_zur_bedarfsdeckung_m: The internal type numpy.ndarray[float] is not yet supported.
E
E           Note that conversion from floating point to integers or Booleans inherently suffers from approximation error. It might well be that your data seemingly obey the restrictions when scrolling through them, but in fact they do not (for example, because 1e-15 is displayed as 0.0).
E           The best solution is to convert all columns to the expected data types yourself.
```

Other functions in the code use `numpy.ndarray[bool]` as expected return and this works. Any idea @lars-reimann?

Comment (lars-reimann) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py]: Looks like GETTSIM raises this error itself. Generally, one should never use floating point numbers to represent exact monetary value, for the reasons cited in the error message. But if a close approximation is fine, we could adjust the error.

Comment (hmgaudecker) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:73]: The problem is that we check for the type of an input Series in `convert_series_to_internal_type`, which needs to be `float`, `bool` etc.:

```
 277             internal_type = None                                                                                                                                            
 278             if column_name in TYPES_INPUT_VARIABLES:                                                                                                                        
 279                 internal_type = TYPES_INPUT_VARIABLES[column_name]                                                                                                          
 280             elif (                                                                                                                                                          
 281                 column_name in functions_overridden                                                                                                                         
 282                 and "return" in functions_overridden[column_name].__annotations__                                                                                           
 283             ):                                                                                                                                                              
 284                 internal_type = functions_overridden[column_name].__annotations__["return"]                                                                                 
 285                                                                                                                                                                             
 286             # Make conversion if necessary                                                                                                                                  
 287             if internal_type and not check_series_has_expected_type(series, internal_type):                                                                                 
 288                 try:                                                                                                                                                        
 289                     data[column_name] = convert_series_to_internal_type(                                                                                                    
 290                         series, internal_type                                                                                                                               
 291                     )                                                                                                                                                       
 292                     collected_conversions.append(                                                                                                                           
 293                         f" - {column_name} from {series.dtype} "                                                                                                            
 294                         f"to {internal_type.__name__}"                                                                                                                      
 295                     )                                                                                                                                                       
 296                                                                                                                                                                             
 297                 except ValueError as e:                                                                                                                                     
 298                     breakpoint()                                                                                                                                            
 299  ->                 collected_errors.append(f" - {column_name}: {e}")                                                                                                       
(Pdb++) column_name
'kindergeld_zur_bedarfsdeckung_m'
(Pdb++) internal_type
numpy.ndarray[float]
(Pdb++) check_series_has_expected_type(series, internal_type)
False
(Pdb++) check_series_has_expected_type(series, float)
True
(Pdb++) functions_overridden[column_name].__policy_info__
*** AttributeError: 'function' object has no attribute '__policy_info__'
```

When we use the correct type annotation for a function that is already vectorised, this fails. @lars-reimann, can you think of a sensible solution here? Probably means making a case distinction regarding the definition of `internal_type` on line 284 above, depending on whether the function has the `skip_vectorization` decorator, though I am afraid the last line of my debugging shows we haven't applied it at that point.

> Other functions in the code use `numpy.ndarray[bool]` as expected return and this works.

Most likely we never use them as input columns in tests directly, so `convert_series_to_internal_type` never gets called. Although I have not checked.

@MImmesberger: I think the short-term hack here is to leave the annotation at `float`.

Comment (lars-reimann) [reply src/_gettsim/transfers/arbeitsl_geld_2/kindergelduebertrag.py:73]: The attribute added by the decorator is called `__info__`, not `__policy_info__`. Something like this should allow detecting whether the function is vectorized:

https://github.com/iza-institute-of-labor-economics/gettsim/blob/add6232460ac7addbd7f82ff21f0c3823e6e89a0/src/_gettsim_tests/test_docs.py#L139
